### PR TITLE
feat: piecedirectory cleanup

### PR DESCRIPTION
**Pull request by LexLuthr**

### CODE REVIEW

**Comment by LexLuthr** [comment]
We probably will need to refactor the couchbase tested so they can succeed on circleCI. Even if I enable remote_docker in circleCI, the tests keep failing. One, because multiple tests require to spin up couchbase and thus required ports get blocked. Other reason is a bit weird but I think sleep is not being honored so couchbase connection starts before services are setup correctly. 
Seems to be working locally.

File: extern/boostd-data/ldb/service.go
Comment (dirkmc) [inline extern/boostd-data/ldb/service.go:353]: Let's return nil if the error is a not found error

File: extern/boostd-data/svc/svc_test.go
Comment (dirkmc) [inline extern/boostd-data/svc/svc_test.go:51]: Do these tests take 5 minutes?

File: extern/boostd-data/svc/svc_test.go
Comment (dirkmc) [inline extern/boostd-data/svc/svc_test.go:424]: This is already tested elsewhere, let's try to keep this test to the minimum required to test the remove functionality

File: extern/boostd-data/client/client.go
Comment (dirkmc) [inline extern/boostd-data/client/client.go:32]: I'm not sure that we need the RemovePieceMetadata method - normally we should only ever need to call RemoveDealForPiece.
RemoveIndexes should only be called if something went wrong and the database got into an inconsistent state where we have indexes but not piece information, right?

File: extern/boostd-data/couchbase/db.go
Comment (dirkmc) [inline extern/boostd-data/couchbase/db.go:624]: I think we should still check to make sure that the piece cid in the array matches the piece cid we're trying to remove. So basically just remove these lines (as we do that check below)
```suggestion
```

File: extern/boostd-data/couchbase/db.go
Comment (dirkmc) [inline extern/boostd-data/couchbase/db.go:640]:
```suggestion
```

**Review by dirkmc — COMMENTED** [review]
This looks like a good start 👍 

There's a lot going on in this PR so I'd suggest we break this PR into smaller PRs and review them one-by-one:
1. Firstly let's create a PR that just does the remove operations
2. Let's create a separate PR branched from `nonsense/piece-store` that gets the couchbase tests working with docker. Note that we should try to use docker's go API rather than executing command-line commands from a shell.
3. Let's create a separate PR branched from `nonsense/piece-store` that has unrelated fixes (eg fixing the span context)
4. Let's move the GC implementation to another PR after the first PR has been merged.
5. Let's minimize changes to go.mod in each PR so it's clear why each dependency is changing.

File: extern/boostd-data/svc/svc_test.go
Comment (LexLuthr) [reply extern/boostd-data/svc/svc_test.go:51]: CouchBase tests are taking longer. Not 5 minutes though. Since, I am removing all CB tests from this PR, I will reset all contexts to 5 Seconds.

File: extern/boostd-data/client/client.go
Comment (LexLuthr) [reply extern/boostd-data/client/client.go:32]: I kept all 3 removal method to cover edge cases where one or more DB operations fail.
Example:
Remove last deal -> remove metadata(fails) -> remove indexes (succeeds)
So now we can remove metadata as indexes are already gone but metadata itself was left.

**Comment by LexLuthr** [comment]
- I have "undo" most of the extra changes and this PR not focuses in cleanup API only. 
- All suggested changes were implemented. 

- https://github.com/filecoin-project/boost/pull/968 created for CB testing.

- I will create a new PR for unrelated fixes once this merged so there are no conflicts.

- I am holding off on creating the GC PR till I have clarity on deal UUID vs legacy deal ID as GC needs to remove both legacy and new deals. There are a couple of other points which needs clarification.

@dirkmc Can you review again?

File: extern/boostd-data/couchbase/service.go
Comment (nonsense) [inline extern/boostd-data/couchbase/service.go:307]:
```suggestion
	log.Debugw("handle.remove-deal-for-piece", "piece-cid", pieceCid, "deal-uuid", dealUuid)
```

File: extern/boostd-data/ldb/service.go
Comment (nonsense) [inline extern/boostd-data/ldb/service.go:378]:
```suggestion
	log.Debugw("handle.remove-deal-for-piece", "piece-cid", pieceCid, "deal-uuid", dealUuid)
```

**Review by nonsense — APPROVED** [review]

File: extern/boostd-data/couchbase/db.go
Comment (dirkmc) [inline extern/boostd-data/couchbase/db.go:747]: This block needs to be wrapped in a compare-and-swap. See AddDealForPiece for an example.
You can read more about compare-and-swap at https://docs.couchbase.com/go-sdk/current/howtos/concurrent-document-mutations.html

File: extern/boostd-data/couchbase/service.go
Comment (dirkmc) [inline extern/boostd-data/couchbase/service.go:344]: I'd suggest moving this code to db.go and putting it inside a compare-and-swap

File: extern/boostd-data/go.mod
Comment (dirkmc) [inline extern/boostd-data/go.mod:7]: There are a bunch of changes to versions in go.mod and go.sum. Let's not change any versions unless we need to.

File: extern/boostd-data/svc/svc_test.go
Comment (dirkmc) [inline extern/boostd-data/svc/svc_test.go:473]: Let's also add checks that PiecesContainingMultihash and IsIndexed work as expected

File: extern/boostd-data/go.mod
Comment (LexLuthr) [reply extern/boostd-data/go.mod:7]: I have changed the dependency back but require{} are automatically updated by go mod tidy as per requirements.

File: extern/boostd-data/couchbase/db.go
Comment (dirkmc) [inline extern/boostd-data/couchbase/db.go:681]: Here we should check if the error is a not found error. If so, we should just return nil.

File: extern/boostd-data/couchbase/db.go
Comment (dirkmc) [inline extern/boostd-data/couchbase/db.go:738]: If the key is not found, let's assume another process removed it, and continue (skip this key)

File: extern/boostd-data/couchbase/db.go
Comment (dirkmc) [inline extern/boostd-data/couchbase/db.go:787]: If the err is a not found error, let's just return nil

File: extern/boostd-data/couchbase/service.go
Comment (dirkmc) [reply extern/boostd-data/couchbase/service.go:344]: Marking as unresolved as it doesn't look like the logic has changed

File: extern/boostd-data/couchbase/db.go
Comment (LexLuthr) [reply extern/boostd-data/couchbase/db.go:738]: This is wrapped in a cas retry func() so, putting a continue statement won't work. I can return the original error here and check the error outside the func() and then continue without calling the remove. WDYT?

Comment (dirkmc) [reply extern/boostd-data/couchbase/db.go]: I think it makes more sense to check if the error is a not found error, and in that case return nil
Then withCasRetry will also return nil.
We should also check the error returned from db.pieceOffsets.Remove and if it's a not found error, ignore it.

File: extern/boostd-data/couchbase/db.go
Comment (dirkmc) [inline extern/boostd-data/couchbase/db.go:796]: I think this check should actually be on line 789

File: extern/boostd-data/couchbase/db.go
Comment (dirkmc) [inline extern/boostd-data/couchbase/db.go:835]: This method should only remove one deal right? If so let's call it RemoveDeal

File: extern/boostd-data/ldb/service.go
Comment (dirkmc) [inline extern/boostd-data/ldb/service.go:448]: Let's rename the db method to RemovePieceMetadata to be consistent with the name in the service

File: piecedirectory/piecedirectory.go
Comment (dirkmc) [inline piecedirectory/piecedirectory.go:330]: Let's rename DeleteDealForPiece to RemoveDealForPiece to be consistent with the service

File: extern/boostd-data/couchbase/db.go
Comment (dirkmc) [reply extern/boostd-data/couchbase/db.go]: This is not resolved

File: extern/boostd-data/couchbase/db.go
Comment (dirkmc) [reply extern/boostd-data/couchbase/db.go]: This is not resolved

**Review by dirkmc — APPROVED** [review]

**Pull request merged**
